[UIAsyncTextInput] Keyboard incorrectly autocapitalizes after deleting backwards

whsieh opened the pull request:
#### 9bf259e8d9e0193d6bbe8bd66259bf9d853dd538
<pre>
[UIAsyncTextInput] Keyboard incorrectly autocapitalizes after deleting backwards
<a href="https://bugs.webkit.org/show_bug.cgi?id=267003">https://bugs.webkit.org/show_bug.cgi?id=267003</a>
<a href="https://rdar.apple.com/120381357">rdar://120381357</a>

Reviewed by Megan Gardner.

This bug was caused by the fact that autocorrection context requests with async text input enabled
broke after the changes in 272018@main — in particular, because we don&apos;t pass the `Text` request
flag when converting from `WebKit::DocumentEditingContext` to `WKSETextDocumentContext` in
`-requestTextContextForAutocorrectionWithCompletionHandler:`.

This happened to work in my testing when `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set,
because the logic in `DocumentEditingContext::toPlatformContext` assumes that `Text` is specified,
if `AttributedText` isn&apos;t set. This was an unintentional departure from the shipping behavior, which
checked:

```
    if (options.contains(DocumentEditingContextRequest::Options::AttributedText)) {

        …

    } else if (options.contains(DocumentEditingContextRequest::Options::Text)) {

        …

    }
```

As such, we fix this by making the `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` codepath consistent
with the above logic, and pass in `DocumentEditingContextRequest::Options::Text` when using
`toPlatformContext`, when returning an autocorrection context in
`-requestTextContextForAutocorrectionWithCompletionHandler:`.

* LayoutTests/editing/deleting/ios/no-autoshift-after-deleting-character-in-word-expected.txt: Added.
* LayoutTests/editing/deleting/ios/no-autoshift-after-deleting-character-in-word.html: Added.

Add a layout test to exercise the bug fix.

* Source/WebKit/Shared/DocumentEditingContext.mm:
(WebKit::DocumentEditingContext::toPlatformContext):

Make this only set non-attributed text context strings in the case where the `Text` request flag is
specified, to match shipping behavior.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView requestTextContextForAutocorrectionWithCompletionHandler:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView autocorrectionRectsForString:]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView effectiveTextInputTraits]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):

Drive-by cleanup around API test infrastructure, with async text input enabled:

-   Some of these codepaths were using `UI_ASYNC_TEXT_INPUT` instead of `UI_ASYNC_TEXT_INTERACTION`,
    effectively becoming dead code.

-   A couple of codepaths related to autocorrection context and rects weren&apos;t actually calling into
    the async text input codepaths as intended.

(wrap):
(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):

Canonical link: <a href="https://commits.webkit.org/272594@main">https://commits.webkit.org/272594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/734bd28c865a51fc739f287232b4c137b6fdb1e3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32254 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10989 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34069 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34812 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29220 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33104 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13384 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8198 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28783 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32672 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9300 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28832 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8068 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8240 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28754 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-works-on-frame-src.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36156 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29198 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34343 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8346 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6279 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32208 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9991 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28522 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7531 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8977 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8890 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
```diff
@@ -264,26 +279,40 @@ - (void)extendSelectionToEndOfParagraph
+#if HAVE(UI_ASYNC_TEXT_INTERACTION)
     if (self.hasAsyncTextInput)
-        return static_cast<id<UITextInputTraits_Private>(self.extendedTextInputTraits);
+        return static_cast<id<UITextInputTraits_Private>>(self.extendedTextInputTraits);
```
How did this even compile before??

whsieh replied:
`UI_ASYNC_TEXT_INPUT` should've been `UI_ASYNC_TEXT_INTERACTION`, so this was essentially dead code.

(I audited for other similar mistakes, and found a couple in `WKContentViewInteraction.mm` which are fixed by this patch)

aprotyas reviewed (commented):
LGTM

whsieh committed:
[UIAsyncTextInput] Keyboard incorrectly autocapitalizes after deleting backwards
https://bugs.webkit.org/show_bug.cgi?id=267003
rdar://120381357

Reviewed by Megan Gardner.

This bug was caused by the fact that autocorrection context requests with async text input enabled
broke after the changes in 272018@main — in particular, because we don't pass the `Text` request
flag when converting from `WebKit::DocumentEditingContext` to `WKSETextDocumentContext` in
`-requestTextContextForAutocorrectionWithCompletionHandler:`.

This happened to work in my testing when `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` is set,
because the logic in `DocumentEditingContext::toPlatformContext` assumes that `Text` is specified,
if `AttributedText` isn't set. This was an unintentional departure from the shipping behavior, which
checked:

```
    if (options.contains(DocumentEditingContextRequest::Options::AttributedText)) {

        …

    } else if (options.contains(DocumentEditingContextRequest::Options::Text)) {

        …

    }
```

As such, we fix this by making the `SERVICE_EXTENSIONS_TEXT_INPUT_IS_AVAILABLE` codepath consistent
with the above logic, and pass in `DocumentEditingContextRequest::Options::Text` when using
`toPlatformContext`, when returning an autocorrection context in
`-requestTextContextForAutocorrectionWithCompletionHandler:`.

* LayoutTests/editing/deleting/ios/no-autoshift-after-deleting-character-in-word-expected.txt: Added.
* LayoutTests/editing/deleting/ios/no-autoshift-after-deleting-character-in-word.html: Added.

Add a layout test to exercise the bug fix.

* Source/WebKit/Shared/DocumentEditingContext.mm:
(WebKit::DocumentEditingContext::toPlatformContext):

Make this only set non-attributed text context strings in the case where the `Text` request flag is
specified, to match shipping behavior.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView requestTextContextForAutocorrectionWithCompletionHandler:]):
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView autocorrectionRectsForString:]):
(-[WKWebView autocorrectionContext]):
(-[WKWebView effectiveTextInputTraits]):
(-[WKWebView replaceText:withText:shouldUnderline:completion:]):

Drive-by cleanup around API test infrastructure, with async text input enabled:

-   Some of these codepaths were using `UI_ASYNC_TEXT_INPUT` instead of `UI_ASYNC_TEXT_INTERACTION`,
    effectively becoming dead code.

-   A couple of codepaths related to autocorrection context and rects weren't actually calling into
    the async text input codepaths as intended.

(wrap):
(-[WKWebView synchronouslyAdjustSelectionWithDelta:]):
(-[WKWebView selectTextForContextMenuWithLocationInView:completion:]):
(-[WKWebView handleKeyEvent:completion:]):

Canonical link: https://commits.webkit.org/272594@main
